evaluator/src/main/kotlin/SourceTree.kt

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright (C) 2022 EPAM Systems, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+package org.ossreviewtoolkit.evaluator
+
+import java.io.File
+
+import org.apache.logging.log4j.kotlin.Logging
+
+import org.ossreviewtoolkit.downloader.Downloader
+import org.ossreviewtoolkit.model.Package
+import org.ossreviewtoolkit.model.VcsInfo
+import org.ossreviewtoolkit.model.VcsType
+import org.ossreviewtoolkit.model.config.DownloaderConfiguration
+import org.ossreviewtoolkit.utils.ort.createOrtTempDir
+
+class SourceTree private constructor(
```

Comment:
Do we really need this new class? Can't we simply use the existing `WorkingTree` instead?

Reply:
The reason I added this class will be more obvious when you look at what get's added in the following commits.
That's all IMO evaluator specific logic. Basically helper functions to use for implementing RuleMatchers and / or policy rules.

Reply:
I saw the upcoming changes, but I'm still not convinced. Looks like the helper functions would as well operate on a `WorkingTree`.

Reply:
> I saw the upcoming changes, but I'm still not convinced.

let me just try on more time to convince you:

I believe even if these helper functions could all be implemented in WorkingTree encapsulation does make sense, because

1. It is impossible that API change in the analyzer (working tree) breaks a policy rule implementation. So, the rules API is  
    independent from working tree.
2. The API can be designed to reflect exactly the requirements from the rules. This is exactly what is needed to arrive at easy to read rules. In particular from my experience it's hard to foresee the exact API needs when implementing new policy rules use cases, and my gut feeling is that this encapsulation will make more and more sense the more functions are being added.
3. The functions can be implement based on the working tree, but it is not required. The encapsulation allows
    changing the implementation. For example a file existence check could be changed to not work on the cloned source but
    on the ScanResult, if that contained a full list of all files.
4. The helper functions are basically factored out logic from the rule matchers added to OrtResult. Exposing that logic is 
   important to expose because the logic inside rule matchers is not re-usable. For example the rule matcher `hasFile()` 
  contains logic to find the actual files. If you want to re-use the logic to find the files, the logic cannot be put into `hasFile()`   
  matcher but needs to be exposed somewhere. That somewhere is the `SourceTree` class I've added. I don't see why that somewhere should be `WorkingTree`.

Is any of the above points somewhat more convincing?

BTW.: I'm planning to add matchers for the commit history, and therefore create (not expose) a working tree instance inside the sourcetree class.

Reply:
Anyhow, as I plan to make a couple of further PRs on top, can we just keep it for now and make that decision when that work is done (guess we know more by then)? If it's really not needed then I'll refactor it away.

Reply:
> can we just keep it for now and make that decision when that work is done

I'd actually prefer to get that sorted out now. IMO we had to much of "I need this urgently now so let's merge it"-style of changes recently, and ORT's code base is starting to suffer from suboptimal code design decisions (with only a single use-case in mind).

Reply:
Ok, then let's sort it out. What do you propose to do?

Reply:
> So, the code base wouldn't suffer from it in the long run. Did you get that?

I did get that. But such promises were made in the past, and then that planned refactoring never happened, or only after a *very* long time.

> I prefer a rather iterative approach

Iterative is fine, but iterative should also mean that it's already going into the right direction, and not introducing stuff that gets removed later on (where foreseeable).

What I'd like to discuss first with you and at least also @mnonnenmacher is whether providing access to the source / working tree should really become a "first class citizen" of the evaluator as currently sketched. When I first read about https://github.com/oss-review-toolkit/ort/issues/5621, I was hoping that would be mostly implemented by helper functions implemented in EPAM's `rules.kts` itself, and not so much in ORT upstream.